client/failing_funding_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2022 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"perun.network/go-perun/channel"
+	chtest "perun.network/go-perun/channel/test"
+	"perun.network/go-perun/client"
+	ctest "perun.network/go-perun/client/test"
+	"perun.network/go-perun/wire"
+	"polycry.pt/poly-go/test"
+)
+
+func TestFailingFunding(t *testing.T) {
+	const (
+		timeout      = 1 * time.Second
+		fridaIdx     = 0
+		fredIdx      = 1
+		fridaInitBal = 100
+		fredInitBal  = 50
+	)
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	rng := test.Prng(t)
+
+	setups := NewSetups(rng, []string{"Frida", "Fred"})
+	// Inject the failing funder for Frida.
+	setups[fridaIdx].Funder = FailingFunder{}
+	clients := NewClients(t, rng, setups)
+	frida, fred := clients[fridaIdx], clients[fredIdx]
+	fridaAddr, fredAddr := frida.Identity.Address(), fred.Identity.Address()
+
+	chsFred := make(chan *client.Channel, 1)
+	errsFred := make(chan error)
+	go fred.Handle(
+		ctest.AlwaysAcceptChannelHandler(ctx, fredAddr, chsFred, errsFred),
+		ctest.AlwaysRejectUpdateHandler(ctx, errsFred),
+	)
+
+	// Create the proposal.
+	asset := chtest.NewRandomAsset(rng)
+	initAlloc := channel.NewAllocation(2, asset)
+	initAlloc.SetAssetBalances(asset, []*big.Int{big.NewInt(fridaInitBal), big.NewInt(fredInitBal)})
+	parts := []wire.Address{fridaAddr, fredAddr}
+	prop, err := client.NewLedgerChannelProposal(
+		challengeDuration,
+		fridaAddr,
+		initAlloc,
+		parts,
+	)
+	require.NoError(t, err, "creating ledger channel proposal")
+
+	// Frida sends the proposal.
+	_, err = frida.ProposeChannel(ctx, prop)
+	// We expect a ChannelFunding error.
+	cfErr, ok := err.(*client.ChannelFundingError)
+	require.Truef(t, ok, fmt.Sprintf("expexted ChannelFundingError, got %T", err))
+	require.Nil(t, cfErr.SettleError)
+
+	select {
+	case <-chsFred:
+		t.Fatalf("expected channel creation to fail")
+	case err = <-errsFred:
+	}
+	assert.Truef(t, errors.As(err, &channel.FundingTimeoutError{}),
```

Comment:
Good that you pointed this out because no, it does not receive `ChannelFundingError` atm.
This is because when the funding for the proposee fails (independently whether it is due to its own funder or due to a timeout), this is handled [here](https://github.com/hyperledger-labs/go-perun/blob/3637bbd3bbca57981de1a736a43010a9d487c1c2/client/proposal.go#L273) and at this point, we should react the same and also settle the channel. I will implement this.

Reply:
I now also implemented that the proposee will settle the channel if the funding fails on its side.

When implementing this, I figured out that one actually needs a new context for calling settling when the funding fails due to a timeout which of course, makes things even more complicated.